### PR TITLE
Allow direct inlining of iterators that have up to 10 yields

### PR DESCRIPTION
**Pull request by ronawho**
Previously we only inlined non-zippered iterators that had exactly 1 yield
statement. This limit was to prevent massive code growth since the body will be
inlined at every yield site. However, given the huge performance penalty paid
for not inlining an iterator, it's worthwhile to pay for some code growth.

Use a default limit of 10, that can be set with --inline-iterators-yield-limit.

Additionally, this adds --report-inlined-iterators, which is similar to
--report-optimized-loop-iterators but for direct iterators only. Use this flag
to lock in the yield limit and show when we will and will not inline iterators.

Closes https://github.com/chapel-lang/chapel/issues/6757

### CODE REVIEW

**Comment by ronawho** [comment]
Enabled by https://github.com/chapel-lang/chapel/pull/7253

@daviditen mind reviewing? Probably easiest to review commit at a time.

File: compiler/resolution/lowerIterators.cpp [hunk 1]
```diff
@@ -1246,6 +1246,17 @@ expandIteratorInline(ForLoop* forLoop) {
   Symbol*   ic       = forLoop->iteratorGet()->symbol();
   FnSymbol* iterator = getTheIteratorFn(ic);
 
+  if (fReportInlinedIterators) {
+    ModuleSymbol *mod = toModuleSymbol(iterator->getModule());
+    INT_ASSERT(mod);
+
+    if (developer || mod->modTag == MOD_USER) {
+      printf("Inlined iterator (%s) in module %s (%s:%d)\n", iterator->cname,
```
Comment (daviditen) [inline compiler/resolution/lowerIterators.cpp]: I think the location where the iterator is getting inlined e.g. `forLoop` is as important as the location where the iterator is defined.

**Comment by daviditen** [comment]
Overall looks good.

Have you done any checking on the compiler performance impact of this?  We may need to tune the limit if 10 proves to cause too much generated code bloat.

File: test/functions/iterators/elliot/report/inline-iters-report.default.good [hunk 2]
```diff
@@ -0,0 +1,58 @@
+Inlined iterator (tenYieldIter) in module inline-iters-report (inline-iters-report.chpl:10)
```
Comment (daviditen) [inline test/functions/iterators/elliot/report/inline-iters-report.default.good]: Is the order of these `--report-inlined-iterators` messages guaranteed, or do you need something like a `.prediff` to sort them?

File: compiler/AST/iterator.cpp [hunk 3]
```diff
@@ -1501,7 +1501,7 @@ void lowerIterator(FnSymbol* fn) {
     // Isn't that bad?
     if (singleLoop) {
       if (fReportOptimizedLoopIterators) {
-        ModuleSymbol *mod = toModuleSymbol(fn->defPoint->parentSymbol);
+        ModuleSymbol *mod = toModuleSymbol(fn->getModule());
```
[the remark this answers is not in the document]
Comment (ronawho) [reply compiler/AST/iterator.cpp]: Good catch. Will remove this along with the assert(mod)

File: compiler/resolution/lowerIterators.cpp [hunk 4]
```diff
@@ -1758,13 +1769,14 @@ countYieldsInFn(FnSymbol* fn)
...
+// Iterators can be inlined if they contains between 1 and
+// inline_iter_yield_limit yield statements
 static bool
 canInlineIterator(FnSymbol* iterator) {
-  unsigned count = countYieldsInFn(iterator);
+  int count = countYieldsInFn(iterator);
 
   // count==0 e.g. in users/biesack/test_recursive_iterator.chpl
-  return (count == 1) ? true : false;
+  return (count >= 1 && count <= inline_iter_yield_limit) ? true : false;
```
[the remark this answers is not in the document]
Comment (ronawho) [reply compiler/resolution/lowerIterators.cpp]: will do

File: compiler/resolution/lowerIterators.cpp (hunk 1 above)
Comment (ronawho) [reply compiler/resolution/lowerIterators.cpp]: I don't really disagree, but for now I want to keep the output consistent with `--report-optimized-loop-iterators`.

File: test/functions/iterators/elliot/report/inline-iters-report.default.good (hunk 2 above)
Comment (ronawho) [reply test/functions/iterators/elliot/report/inline-iters-report.default.good]: I was wondering the same thing. `functions/iterators/sungeun/report/` (which tests `--report-optimized-loop-iterators`) does not sort and it has been stable for the past 4 years, so I think the ordering is relatively safe. Note that the `--inline-iterators` and `--optimize-loop-iterators` functions are both called from `expandForLoop()`, so this is no less stable than that test.

**Comment by ronawho** [comment]
>Have you done any checking on the compiler performance impact of this? We may need to tune the limit if 10 proves to cause too much generated code bloat. 

I did not see any compiler performance impact, but we don't have many tests for iters with multiple yields. And our core range/domain/array iterators have all been optimized to only have a single yield, so we wouldn't see an impact from our core modules.

I agree that we may have to tune in the future, but at least for now a max of 10x growth for an iterator doesn't seem too horrendous given how large our generated code is anyways.

**Comment by daviditen** [comment]
OK, that works for me.

**Pull request merged**